### ISSUE TITLE
Fails to read license.txt

### ISSUE THREAD

**Issue by ronpandolfi** [vispy/vispy#1330]
In python 3.6 (not in 2.7), importing vispy.scene produces the following error:
```
from vispy import scene
Traceback (most recent call last):
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2847, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-3b82c3c73018>", line 1, in <module>
    from vispy import scene
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/scene/__init__.py", line 35, in <module>
    from .visuals import *  # noqa
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/scene/visuals.py", line 17, in <module>
    from .. import visuals
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/visuals/__init__.py", line 14, in <module>
    from .cube import CubeVisual  # noqa
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/visuals/cube.py", line 8, in <module>
    from ..gloo import set_state
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/gloo/__init__.py", line 47, in <module>
    from . import gl  # noqa
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/gloo/gl/__init__.py", line 32, in <module>
    from ._constants import *  # noqa
  File "/opt/pycharm-community/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site-packages/vispy/gloo/gl/_constants.py", line 326, in <module>
    if repr(ob).startswith('GL_'):
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site.py", line 443, in __repr__
    self.__setup()
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/site.py", line 430, in __setup
    data = fp.read()
  File "/home/rp/PycharmProjects/xi-cam/venv3/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 0: ordinal not in range(128)

```
After inspection, it looks like its trying to read its LICENSE.txt, but fails with an encoding error.

One solution would be to change line 429 to `fp = open(filename, "rU", encoding='utf-8')`

**Comment by davidh-ssec** [comment]
What distribution are you running on? Could you please run the following to tell us your default encoding:

    import sys
    print(sys.stdout.encoding)

I'm ok forcing the encoding when loading, but I use vispy on mac, windows, and linux with python 3.6 so I'm curious what's different about your environment that is causing this.

**Comment by davidh-ssec** [comment]
Also have you tried running this outside of pycharm? Or in pycharm but without the debugger (it looks like you are using the debug mode)?

**Comment by ronpandolfi** [comment]
The sys encoding is ANSI_X3.4-1968. I'm on Arch linux, and get the same
result outside of pycharm and debug mode.

**Comment by davidh-ssec** [comment]
@ronpandolfi I'm having trouble reproducing this in an Arch linux VM using Anaconda with python 3.6. After looking at your traceback again I'm not sure this is an issue with vispy. Have you used this environment for other python code? The `_constants.py` module is performing a `for` loop over all of the global variables in itself which includes the special `__builtins__` dictionary. This dictionary (and the interpreter itself) contains a `copyright` variable holding the copyright text for python. It would seem that your python environment was installed or includes modified copyright text.

Could you please try running the following from that environment:

    print(repr(copyright))

Do you get an exception? What about:

    print(repr(__builtins__.__dict__))

**Comment by kmuehlbauer** [comment]
@ronpandolfi If this is still an issue for you, please reopen with added information.

**Comment by Jonseed** [comment]
I'm seeing this bug with:

`UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 2575: character maps to <undefined>`

I don't get a traceback with `print(repr(copyright))` but I do get a traceback with print(repr(__builtins__.__dict__))`

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\x\.conda\envs\3DP\lib\_sitebuiltins.py", line 61, in __repr__
    self.__setup()
  File "C:\Users\x\.conda\envs\3DP\lib\_sitebuiltins.py", line 51, in __setup
    data = fp.read()
  File "C:\Users\x\.conda\envs\3DP\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 2575: character maps to <undefined>
```

This is a brand new Python 3.7.13 conda environment. Is there anything that can be done to fix the Python installation/environment?

**Comment by djhoese** [comment]
@Jonseed As mentioned above, this is not an issue with vispy and is something wrong between your system/environment and python itself (or whatever IDE/interpreter/shell you are using). I can try to help you figure out what you've done that is causing this, but every time we ask someone for more information we get silence. I assume something was figured out either on their own or in some other forum (stackoverflow?).

**Comment by haesleinhuepf** [comment]
I'm having a similar issue when starting napari:
```
(bio39) C:\structure\code\napari-noise2self-adhoc>napari
Traceback (most recent call last):
  File "C:\Users\rober\miniconda3\envs\bio39\Scripts\napari-script.py", line 5, in <module>
    from napari.__main__ import main
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\napari\__main__.py", line 16, in <module>
    from .utils.translations import trans
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\napari\utils\__init__.py", line 2, in <module>
    from .colormaps import Colormap
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\napari\utils\colormaps\__init__.py", line 2, in <module>
    from .colormap import Colormap
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\napari\utils\colormaps\colormap.py", line 7, in <module>
    from napari.utils.color import ColorArray
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\napari\utils\color.py", line 7, in <module>
    from napari.utils.colormaps.standardize_color import transform_color
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\napari\utils\colormaps\standardize_color.py", line 27, in <module>
    from vispy.color import ColorArray, get_color_dict, get_color_names
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\vispy\color\__init__.py", line 12, in <module>
    from .colormap import (Colormap, BaseColormap,  # noqa
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\vispy\color\colormap.py", line 14, in <module>
    import vispy.gloo
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\vispy\gloo\__init__.py", line 47, in <module>
    from . import gl  # noqa
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\vispy\gloo\gl\__init__.py", line 34, in <module>
    from ._constants import *  # noqa
  File "C:\Users\rober\miniconda3\envs\bio39\lib\site-packages\vispy\gloo\gl\_constants.py", line 330, in <module>
    if repr(ob).startswith('GL_'):
  File "C:\Users\rober\miniconda3\envs\bio39\lib\_sitebuiltins.py", line 61, in __repr__
    self.__setup()
  File "C:\Users\rober\miniconda3\envs\bio39\lib\_sitebuiltins.py", line 51, in __setup
    data = fp.read()
  File "C:\Users\rober\miniconda3\envs\bio39\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 2575: character maps to <undefined>
```

Can anyone give a hint of which file is being read where the encoding might be wrong?

**Comment by haesleinhuepf** [comment]
I managed to fix this issue by [changing this file](https://github.com/vispy/vispy/blob/c52c48194c0af60ce97ac844eb87a54f8434988c/vispy/gloo/gl/_constants.py#L329)

from: 
```
ENUM_MAP = {}
for ob in list(globals().values()):
    if repr(ob).startswith('GL_'):
        ENUM_MAP[int(ob)] = ob
del ob
```

to:
```
ENUM_MAP = {}
for ob in list(globals().values()):
    try:
        if repr(ob).startswith('GL_'):
            ENUM_MAP[int(ob)] = ob
    except:
        pass
del ob
```

Let me know @djhoese if that's a good idea and I'll send a PR.

Thanks!

**Comment by djhoese** [comment]
So in #2386 which appears to be a duplicate of this I discovered that this is actually a bug in CPython that was fixed in Python 3.10:

https://github.com/vispy/vispy/issues/2386#issuecomment-1315274141

As for your fix, that's a quick workaround but after investigating what the code is actually trying to do when debugging in #2386, I think a better solution might be the following:

```python
ENUM_MAP = {}
for var_name, ob in list(globals().items()):
    if var_name.startswith('GL_'):
        ENUM_MAP[int(ob)] = ob
del ob, var_name
```

This has the added benefit that we don't even try `repr`'ing the objects at all. If you look at all the constants defined in that module they all start with `GL_` so this just avoids dealing with anything else. So if you're willing to submit a PR with this fix @haesleinhuepf it would be much appreciated. Note that the edit would have to occur here:

https://github.com/vispy/vispy/blob/c52c48194c0af60ce97ac844eb87a54f8434988c/codegen/createglapi.py#L117-L123

As the `_constants.py` module is generated from this code. I think I will then need to run the code to regenerate the files.
